src/sagas/index.js

```diff
@@ -143,85 +143,115 @@ export function* transformAndPermitVm (vm) {
-  }
-
-  function* currentPoolsIds ({ payload: { count, page } }) {
-    const start = count * page
-    const end = start + count
-    return Array.from(yield select(state => state.vms.get('pools').keys())).slice(start, end)
-  }
-
   //
-  // If more pages are expected, fetch the next page and grab the ids fetched
-  // If no more pages are expected, grab the current page of ids from the redux store
+  // If more pages are expected, fetch the next page
+  // If no more pages are expected, skip the fetch
   //
   const count = AppConfiguration.pageLimit
-  const [vms, pools] = yield all([
-    call(vmsExpectMorePages ? fetchVms : currentVmsIds, { payload: { count, page: vmsPage + 1 } }),
-    call(poolsExpectMorePages ? fetchPools : currentPoolsIds, { payload: { count, page: poolsPage + 1 } }),
-  ])
+  const {
+    vms: { internalVms: vms = null },
+    pools: { internalPools: pools = null },
+  } = yield all({
+    vms:
+      vmsExpectMorePages
+        ? call(fetchVms, { payload: { count, page: vmsPage + 1 } })
+        : {},
+
+    pools:
+      poolsExpectMorePages
+        ? call(fetchPools, { payload: { count, page: poolsPage + 1 } })
+        : {},
+  })
+
+  // TODO: move to reducer, keep state changes in 1 action...
+  // Put the new page of data to the store
+  if (vms) {
+    yield put(updateVms({ vms, copySubResources: true }))
+  }
+  if (pools) {
+    yield put(updatePools({ pools }))
+  }
+  yield put(updateVmsPoolsCount())
 
   //
   // Since the REST API doesn't give a record count in paginated responses, we have
   // to guess if there is more to fetch.  Assume there is more to fetch if the page
   // of ids fetched/accessed is full.
   //
   yield put(updatePagingData({
-    vmsPage: vmsExpectMorePages ? vmsPage + 1 : undefined,
-    vmsExpectMorePages: vms.length >= count,
-    poolsPage: poolsExpectMorePages ? poolsPage + 1 : undefined,
-    poolsExpectMorePages: pools.length >= count,
+    ...vmsExpectMorePages
+      ? {
+        vmsPage: vmsPage + 1,
+        vmsExpectMorePages: vms ? vms.length >= count : false,
+      }
+      : {
+        vmsExpectMorePages: vmsCount >= (vmsPage * count),
+      },
+
+    ...poolsExpectMorePages
+      ? {
+        poolsPage: poolsPage + 1,
+        poolsExpectMorePages: pools ? pools.length >= count : false,
+      }
+      : {
+        poolsExpectMorePages: poolsCount >= (poolsPage * count),
+      },
   }))
+  // TODO: ...move to reducer, keep state changes in 1 action
+
+  if (vms) {
+    yield fetchUnknownIcons({ vms })
+  }
 }
 
 export function* fetchVms ({ payload: { count, page, shallowFetch = true } }) {
-  const fetchedVmIds = []
-
   const additional = shallowFetch ? VM_FETCH_ADDITIONAL_SHALLOW : VM_FETCH_ADDITIONAL_DEEP
   const apiVms = yield callExternalAction('getVms', Api.getVms, { payload: { count, page, additional } })
-  if (apiVms && apiVms.vm) {
-    const internalVms = []
+
+  let internalVms = null
```

Comment:
`null` breaks default values - consider different alternatives

Reply:
It doesn't break anything if null is the intended value.

(Had to double check how destructuring works with defaults ... the default is selected if and only if the value is `undefined`, not truthy false)